File: src/db-gwent-server-kotlin/src/main/kotlin/entities/Game.kt
package entities

class Game(
    val id: GameId,
    val players: List<Player>,
    val turn: Turn = Turn(players)
) {
    fun playCard(playerId: PlayerId, cardId: CardId, rowType: RowType? = null) =
        turn.run(playerId) {
            players.getOfId(playerId)?.playCard(cardId) ?: false
        }
}

data class GameId(val value: String)

fun List<Player>.getOfId(id: PlayerId) = this.find { it.id == id }

enum class RowType {
    Any,
    Close,
    Ranged,
    CloseOrRanged,
    Siege;
}

class Turn(
    val players: List<Player>,
    private var nextToPlay: PlayerId? = null) {

    fun run(playerId: PlayerId, action: () -> Boolean) =
        canPlay(playerId) && action() && setPlayed(playerId)

    private fun canPlay(playerId: PlayerId) =
        players.getOfId(playerId) != null && nextToPlay == playerId

    private fun setPlayed(playerId: PlayerId) =
        playerId != nextToPlay && go { nextToPlay = playerId }
}

fun go(action: () -> Unit) = action().run { true }

File: src/db-gwent-server-scala/src/main/scala/com/entities/Game.scala
package com.entities

import com.basic.{getOfId, getNotOfId, ifOkThen, t}

case class GameId(value: String)
case class Game(
  id: GameId, players: Vector[Player], turn: Turn):
  def this(id: GameId, players: Vector[Player]) =
    this(id, players, Turn(players.map(p => p.id).toList, players.head.id))

  def playCard(playerId: PlayerId, cardId: CardId, row: TargetRowType = null): Boolean =
    turn.run(playerId): () =>
      players
        .getOfId(playerId) match
          case Some(player) =>
            getEnemyAndPlayCard(player, cardId, row)
            toNextRound(player, cardId)
          case None => false

  private def toNextRound(player: Player, cardId: CardId): Boolean =
    (player.getCard(cardId), players.getNotOfId(player.id)) match
      case (Some(card), Some(otherPlayer)) =>
        card.abilities.forall(a => a.apply(card, player, otherPlayer))
      case _ => false

  private def getEnemyAndPlayCard(player: Player, cardId: CardId, row: TargetRowType = null): Boolean =
    players.getNotOfId(player.id) match
      case Some(otherPlayer) => player.playCard(cardId, row, otherPlayer)
      case _ => false